cEP-0028.md

```diff
@@ -0,0 +1,495 @@
+       Resolve tries to find an available proxy or creates one
+       if there is no available proxy for the said file.
+
+       :param filename:
+           The filename to search for in the map or to create
+           a proxy instance using.
+       :param workspace:
+           Used in case the lookup fails and a new instance is
+           being initialized.
+       :hard_sync:
+           Boolean flag indicating if the file should be initialized
+           from the file on disk or fail otherwise.
+       :return:
+           Returns a proxy instance or raises associated exceptions.
+       """
+       proxy = self.get(filename)
+       if proxy is not None:
+         return proxy
+
+       try:
+         proxy = FileProxy.from_name(filename, workspace)
+       except OSError as exception:
+         if hard_sync:
+           raise exception
+
+       # possibly raises FileNameError exception
+       proxy = FileProxy(filename, workspace)
+       self.add(proxy)
+
+       return proxy
+   ```
+
+   - `coalib.processes.Processing.get_file_dict()`
+
+     Requires patching `get_file_dict()` to conditionally build
+     file dict from a FileProxyMap instance if available rather
+     than from disk files.
+
+   ```python
+   def get_file_dict(filename_list,
+                     log_printer=None,
+                     allow_raw_files=False,
+                     fileproxy_map=None):
+     """
+     Reads all files into a dictionary.
+     :param filename_list:   List of files to get the contents of.
+     :param log_printer:     The logger which logs errors.
+     :param allow_raw_files: Allow the usage of raw files (non text files),
+                             disabled by default
+     :param fileproxy_map:   The FileProxyMap instance to use while building
+                             a file dict when available.
+     :return:                Reads the content of each file into a dictionary
+                             with filenames as keys.
+     """
+     file_dict = {}
+     for filename in filename_list:
+       try:
+         if fileproxy_map is not None:
+           proxy = fileproxy_map.resolve(filename,
+                                         hard_sync=True)
+           file_contents = proxy.contents()
+           file_lines = file_contents.splitlines(keepends=True)
+           file_dict[filename] = tuple(file_lines)
+         else:
+           with open(filename, 'r',
+                     encoding=detect_encoding(filename)) as _file:
+                     file_dict[filename] = tuple(_file.readlines())
+       except UnicodeDecodeError:
+         if allow_raw_files:
+           file_dict[filename] = None
+           continue
+
+         logging.warning("Failed to read file '{}'. It seems to contain "
+                         'non-unicode characters. Leaving it out.'
+                         .format(filename))
+       except (OSError, FileNameError) as exception:
+         if fileproxy_map is not None and allow_raw_files:
```

Comment:
this is ok for your prototype in the cEP, however your exclusion of `allow_raw_files` here suggests there are bigger problems when `allow_raw_files` and `fileproxy_map` are used together.  Dont hide the problem.  When you implement this, detect this scenarion as early as possible and raise `NotImplementedError`

Reply:
Yeah, FileProxies were primarily built with text files and code editors in mind, although I do not see how they fail to be compatible with raw files, I haven't tested them. I will remember to fix this issue.